src/dotnet/OpenAPI.Validator/Validation/ArmResourcePropertiesBag.cs

```diff
@@ -47,19 +48,64 @@ public class ArmResourcePropertiesBag : TypedRule<Dictionary<string, Schema>>
         {
-            var resModels = context.ResourceModels;
-            var violatingModels = resModels.Where(res => definitions[res].Properties?.ContainsKey("properties") == true
-                                                         && definitions[res].Properties["properties"]?.Properties?.Keys.Intersect(ArmPropertiesBag).Any() == true);
-            foreach (var violatingModel in violatingModels)
+            IEnumerable<string> resourceModelsWithPropertiesBag = context.ResourceModels.Where(resourceModel => definitions[resourceModel].Properties?.ContainsKey("properties") == true);
+            foreach(string resourceModelName in resourceModelsWithPropertiesBag)
             {
-                var violatingProperties = definitions[violatingModel].Properties.Keys.Union(definitions[violatingModel].Properties["properties"].Properties.Keys).Intersect(ArmPropertiesBag);
-                yield return new ValidationMessage(new FileObjectPath(context.File, context.Path.AppendProperty(violatingModel).AppendProperty("properties")), this, violatingModel,
-                                                   string.Join(", ", violatingProperties));
+                Dictionary<string, IEnumerable<string>> violations = new Dictionary<string, IEnumerable<string>>();             
+                CheckModelForViolation(definitions[resourceModelName], resourceModelName, definitions, context, violations);
+
+                foreach (string modelName in violations.Keys)
+                {
+                    yield return new ValidationMessage(new FileObjectPath(context.File,
+                    context.Path.AppendProperty(modelName).AppendProperty("properties")), this, modelName,
+                                                   string.Join(", ", violations[modelName]));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Method to check the model for violation of this rule.
+        /// This method checks if the model has propeties bag. If so and it
+        /// has reference then the reference is processed. Else, the individual
+        /// properties are checked for violation.
+        ///
+        /// Then, if the model has allOf's, they need to be processed individually.
+        /// </summary>
+        /// <param name="resourceModel">schema of the model</param>
+        /// <param name="resourceModelName">name of the model</param>
+        /// <param name="definitions">definitions</param>
+        /// <param name="context">context</param>
+        /// <param name="violations">violations table</param>
+        private void CheckModelForViolation(Schema resourceModel, string resourceModelName, Dictionary<string, Schema> definitions, RuleContext context, Dictionary<string, IEnumerable<string>> violations)
+        {
+            if (resourceModel.Properties?.ContainsKey("properties") == true)
```

Comment:
If i am not mistaken then method `Schema.FindReferencedSchema` returns `null` [here](https://github.com/Azure/azure-openapi-validator/blob/42ae8eedab0a3894443966e4513fd28ae5aa77d8/src/dotnet/OpenAPI.Validator/Model/Schema.cs#L87). We are calling that method at line `108`'s first parameter without checking it for `null`. In that case we need to do a null check here.

In general, I'd always prefer to do a null check for all the arguments of any input function as we don't know how someone is going to call it. To be on the safe side.